### ISSUE TITLE
Allow Services to select only the latest backend Pods

### ISSUE THREAD

**Issue by sielaq** [kubernetes/kubernetes#129159]
### What would you like to be added?

Add a new field `spec.selectorLatest` to the `kind: Service` resource to enable automatic selection of the latest pods without need of sophisticated controller to make updates. 

So side to pod selector, would be extra `selectorLatest: true` like

```
spec:
  selectorLatest: true
  selector:
    service_name: foo
```

This will ensure that Load balancer contains ONLY the latest new version, in case of Rolling update.
Which means, depends on rollingUpdate strategy and race conditions, it might have just one or more instance - but this should be enough to serve a new version. 

## Why is needed

In many dynamic environments, managing Service selectors with controller can be cumbersome. When a Deployment creates a new ReplicaSet, the Service's `spec.selector` must be updated rapidly if changes occur. Automating this process via a `selectorLatest: true` field could simplify operations and reduce errors (race conditions).

## Additional Context

This feature would reduce the operational overhead for users running applications in Kubernetes, 
 particularly for CI/CD pipelines and dynamic workloads.

## Example 

Having two `kind: service` exposed publicly
* one traditional - serving dynamic content ( serving old and new version while deploying ) - heavy traffic 
* one via caching CDN with new feature - serving uncached static files only (different domain) - low traffic

ASCII art: 🥲 
```
   www.example.com [   dynamic content ]  -> [kind: Service traditional]              -> same_app
static.example.com [CDN - cache statics]  -> [kind: Service with latest version true] -> same_app
```

The dynamic content could be served in parallel with static content from the same app and using rolling update strategy.

## Benefits

* Simple unified deployment; No need for sophisticated A/B deployment
* Simple CI/CD; No need to warm up caches upfront ( can be triggered during/after deployment or not at all )
* No race conditions in serving static files - always first request for new static can be served by first instance in Service and then cached in CDN - no heavy load on first instance thx to cache upfront.

**Comment by sftim** [comment]
Can we solve this by adding a new label to ReplicaSets and other API kinds that manage Pods? If the newest ReplicaSet had a well-known label on it, managed by the control plane, we wouldn't need to change any stable APIs to get the outcome that @sielaq wants.

(It might still need a KEP)

**Comment by sftim** [comment]
/retitle Allow Services to select only the latest backend Pods

**Comment by aojea** [comment]
I don't think we are going to build this complexity in Services , but I think Gateway API has something on this area

@robscott @shaneutt @youngnick

**Comment by thockin** [comment]
There's not enough information here to assess fully.  In particular, I don't know what "latest" means - are you assuming a standard label which has an integer value or something?

Suppose I have a Deployment with 1000 replicas, and a service pointing to it.  I want to roll out a new version - what do you envision happening?

**Comment by sielaq** [comment]
Hi @thockin

> what do you envision happening?

When using the `rollingUpdate` strategy, you typically set an extra `maxSurge` to allow x% over-deployment. For example, if `maxSurge` is set to 50% and you have 1000 replicas of version `v1`, an additional 500 pods of version `v2` will be spawned.

At the moment the first `v2` pod is running (this depends on whether a readinessProbe is defined and kicks in; if not, it happens when the pod is considered healthy) and is added to the load balancer (`kind: service`), the only version available in the service will be `v2`.

This means the first `v2` pod to appear in the service effectively replaces version `v1`, even if it's just one pod from the 500. The remaining `v2` pods will likely follow shortly thereafter and also join the service.

> are you assuming a standard label which has an integer value or something? 
No, not a label. I’m assuming that only the latest ReplicaSet will be included in the `kind: service` once its first pod is considered healthy.

I hope this explanation makes the idea clearer

regards
Wojciech

**Comment by thockin** [comment]
> This means the first v2 pod to appear in the service effectively replaces version v1

That sound disastrous for large-N services.  If v1 was provisioned to need 1000 replicas, flipping it to v2 after the first replica will just crush it.  This can very likely lead the deployment controller to stop the rolling update because the pods are unhealthy.

>> are you assuming a standard label which has an integer value or something?
>
> No, not a label. I’m assuming that only the latest ReplicaSet will be included in the kind: service once its first pod is considered healthy.

What does "latest" mean?  The one with the newest creationTimestamp?

**Comment by sielaq** [comment]
> That sound disastrous for large-N services

Not at all. You’re forgetting that the primary traffic for the 1000 replicas flows through the traditional service:
```
   www.example.com [   dynamic content ]    -> [ kind: Service traditional ]                 -> same_app
static.example.com [ CDN - cache statics ]  -> [ kind: Service with `selectorLatest: true` ] -> same_app
```
The traffic you’d see on this special service consists of only a few requests from the CDN for new static files - everything else is already cached, everything new will be cached too. 

> What does "latest" mean? The one with the newest creationTimestamp?

In my opinion, the best fit would be to use the `kind: ReplicaSet` as the source of truth, relying on its `resourceVersion`.
Using a pod’s `creationTimestamp` is not a good choice because it changes with every pod crash and is unique to each pod.
However, you might have a better idea to approach this.

**Comment by aojea** [comment]
See if this matches your expectactions https://gateway-api.sigs.k8s.io/guides/traffic-splitting

**Comment by sielaq** [comment]
@aojea unfortunately not, since `traffic splitting` is for different scenarios and it is based on different service names. In my scenario I have one same app and two ways of accessing it. I'm not splitting traffic, I'm "merging" - but I need one uniq way.

**Comment by aojea** [comment]
but you are merging it because you are deciding to use the same deployment, right?
if you use two deployments you'll achieve what want, no?

**Comment by sielaq** [comment]
Unfortunately, it’s not that simple.

If one app creates two deployments simultaneously, I’ll encounter a race condition.
In that scenario, the "traditional-service" might start serving a version with new static files, while those static files haven’t been deployed yet.

Additionally, deploying the two deployments of same app, one after another, isn’t supported by our simple pipeline, which further complicates this approach.

**Comment by thockin** [comment]
> The traffic you’d see on this special service consists of only a few requests from the CDN for new static files - everything else is already cached, everything new will be cached too.

You are describing a VERY specific usage pattern thathappens to be what you need, but I don't think is generally true.  It is not a safe assumption that we can flip all traffic as soon as the first replica of the new version comes up.

What you have are two services, and the "safe" way to do rollouts is either to blend them (ramp one up while the other ramps down) or to bring up the WHOLE new deployment and switch, either atomically or incrementally.  We don't have a traffic-splitting model in Service (better served by Gateway, I think), but an atomic switch is effectively "change the service's selector" (though nothing is really atomic).

I am missing why that's not appropriate for you?

* Deploy app v1 with labels "app: myapp" and "version: v1".
* Service "www" selects "app: myapp"
* Service "static" selelcts "app: myapp, version: v1"
* Start a rollout of app v2 with labels "app: myapp" and "version: v2".
* At some point you decide that it is "far enough"
* Change service "static" to selelct "app: myapp, version: v2"
* Finish rollout
